Automatically launch browser and wait for login; remove external tunnel;

cwm9cwm9 opened the pull request:
**This is a minor rewrite offered for your consideration.**

**Changes**:

1. Removed the ngrok service (because running a major financial access key through a non-open-source uncontrolled 3rd party server makes me uncomfortable) and replaced it with a localhost HTTPS server.  (No tunnel is needed anyway --- the local browser has access to the local computer.)
2. Calling init now only starts the HTTPS server.
3. Calling token will automatically display instructions and launch the browser if no token is currently available, or if the existing token is stale.  The program will wait until the user logs in, then get the key and continue.

NOTES:
_You MUST provide a .crt and .key file for the HTTPS server to run properly.  Install OpenSSL (or equivalent) if you do not already an sn SSL implementation._

    openssl genrsa 2048 > server.key
    chmod 400 host.key # not needed on windows
    openssl req -new -x509 -nodes -sha256 -key server.key -out server.crt

You should always call init at the start of your program.  Call getToken when you need it; the browser will automatically launch if required:

    await init('RDXO55TZCJS4LGIEOF5MDRHMCZVNSFP8B')  // replace with your APP_ID
        .then(console.log("Token server started"))
        .catch((err) => console.log(err));

Then, when you need to use a token:

    TDtoken = await getToken('RDXO55TZCJS4LGIEOF5MDRHMCZVNSFP8B')  // replace with your APP_ID
        .then(console.log("Got token"))
        .catch((err) => console.log(err));

distrill commented on src/init.js:
```diff
@@ -1,27 +1,53 @@
-const { promises: fs } = require('fs');
+var fs = require('fs');
+const { promises: pfs } = require('fs');
+var https = require('https');
```
Can we use const (or let, if needed)?

cwm9cwm9 replied:
Yes.  I apologize, I'm started programming 40 years ago, and this is my first time coding in Java, so I'm a little unfamiliar with the language and conversions.  I don't see why const can't be used everywhere here.  Changed.

distrill commented on src/init.js:
```diff
@@ -1,27 +1,53 @@
+var privateKey = fs.readFileSync('server.key', 'utf8');
+var certificate = fs.readFileSync('server.crt', 'utf8');
+var credentials = { key: privateKey, cert: certificate };
 const ngrok = require('ngrok');
```
This was left in by mistake? I used yarn in this project, `yarn remove ngrok` and `yarn add open` should suffice?

cwm9cwm9 replied:
Yes, I did leave the ngrok package reference in the code by mistake.  Removed.

I am unfamiliar with yarn, but I installed and executed your suggested commands:

    yarn remove ngrok
    yarn add open

but got

    error eslint@6.8.0: The engine "node" is incompatible with this module. Expected version "^8.10.0 || ^10.13.0 || >=11.10.1". Got "11.6.0"
    error Found incompatible module.

I then ran

    yarn remove ngrok --ignore-engines
    yarn add open --ignore-engines

and everything seemed fine.

I committed the changes to package.json and yarn.lock that resulted.

Hope I did that correctly!

distrill reviewed (changes requested):
This is a great improvement, thanks for the contribution. I have a couple of style nitpicks I'd like to address before this gets merged, and then I'll update the npm version.

I also wonder if you'd be willing to update the readme so that it contains the walkthrough in your description?

cwm9cwm9 commented:
> This is a great improvement, thanks for the contribution. I have a couple of style nitpicks I'd like to address before this gets merged, and then I'll update the npm version.
> 
> I also wonder if you'd be willing to update the readme so that it contains the walkthrough in your description?

I have changed the readme but didn't look over it very carefully.

One thing to note is that the CLI interface doesn't make any sense now: running init only starts the HTTPS server, it doesn't actually launch the login.  I removed that reference.  I don't know if you want to do anything about this, but I didn't really see a need to keep the CLI version in.  I suppose a CLI call could be written that inits and then immediately retrieves a token to permit CLI usage.

LMK what you want to do.

distrill commented:
This is great, thanks. I'll merge as is, although I don't think the CLI is particularly useful so I'll probably just get rid of that before too long.
Thanks for the contribution 👍
